xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,364 @@
+	ctrl1, ok1, cancelWatch1 := watchAndFetchNewController(t, client, resourceName, ctrlCh)
+	if !ok1 {
+		t.Fatalf("want a new controller to be build, got none")
+	}
+
+	var cancelWatch2 func()
+	// Call the watch with the same authority name. This shouldn't create a new
+	// controller.
+	resourceNameSameAuthority := buildResourceName(xdsresource.ClusterResource, testAuthority, testCDSName+"1", nil)
+	ctrl2, ok2, cancelWatch2 := watchAndFetchNewController(t, client, resourceNameSameAuthority, ctrlCh)
+	if ok2 {
+		t.Fatalf("an unexpected controller is built with config: %v", ctrl2.config)
+	}
+
+	cancelWatch1()
+	if ctrl1.done.HasFired() {
+		t.Fatalf("controller is closed immediately when the watch is canceled, wanted to be put in the idle cache")
+	}
+
+	// Cancel the second watch, should put controller in the idle cache.
+	cancelWatch2()
+	if ctrl1.done.HasFired() {
+		t.Fatalf("controller is closed when the second watch is closed")
+	}
+
+	time.Sleep(idleTimeout * 2)
+	if !ctrl1.done.HasFired() {
+		t.Fatalf("controller is not closed after idle timeout")
+	}
+}
+
+// TestAuthorityClientClose covers that the authorities in use and in idle cache
+// are all closed when the client is closed.
+func (s) TestAuthorityClientClose(t *testing.T) {
+	overrideFedEnvVar(t)
+
+	ctrlCh, cleanupController := overrideNewController()
+	t.Cleanup(cleanupController)
+
+	client, err := newWithConfig(&bootstrap.Config{
+		XDSServer: serverConfigs[0],
+		Authorities: map[string]*bootstrap.Authority{
+			testAuthority: {XDSServer: serverConfigs[1]},
+		},
+	}, defaultWatchExpiryTimeout, defaultIdleAuthorityDeleteTimeout)
+	if err != nil {
+		t.Fatalf("failed to create client: %v", err)
+	}
+	t.Cleanup(client.Close)
+
+	resourceName := testCDSName
+	ctrl1, ok1, cancelWatch1 := watchAndFetchNewController(t, client, resourceName, ctrlCh)
+	if !ok1 {
+		t.Fatalf("want a new controller to be build, got none")
+	}
+
+	resourceNameWithAuthority := buildResourceName(xdsresource.ClusterResource, testAuthority, testCDSName, nil)
+	ctrl2, ok2, _ := watchAndFetchNewController(t, client, resourceNameWithAuthority, ctrlCh)
+	if !ok2 {
+		t.Fatalf("want a new controller to be build, got none")
+	}
+
+	cancelWatch1()
+	if ctrl1.done.HasFired() {
+		t.Fatalf("controller is closed immediately when the watch is canceled, wanted to be put in the idle cache")
+	}
+
+	// Close the client while watch2 is not canceled. ctrl1 is in the idle
+	// cache, ctrl2 is in use. Both should be closed.
+	client.Close()
+
+	if !ctrl1.done.HasFired() {
+		t.Fatalf("controller in idle cache is not closed after client is closed")
+	}
+	if !ctrl2.done.HasFired() {
+		t.Fatalf("controller in use is not closed after client is closed")
+	}
+}
+
+// TestAuthorityRevive covers that the authorities in the idle cache is revived
+// when a new watch is started on this authority.
+func (s) TestAuthorityRevive(t *testing.T) {
```

Comment:
Can this test be combined with the idleness test, because they are mostly doing the same thing.